tools/generate_report.py

```diff
@@ -0,0 +1,206 @@
+#############################################################################
+##
+##  This file is part of GAP, a system for computational discrete algebra.
+##
+##  Copyright of GAP belongs to its developers, whose names are too numerous
+##  to list here. Please refer to the COPYRIGHT file for details.
+##
+##  SPDX-License-Identifier: GPL-2.0-or-later
+##
+
+# Run this script on main branch with data and gh-pages worktree
+
+"""
+This script compares the current test-status.json with a previous version,
+generates a main report.md along with
+    a test-status-diff.json,
+    a html-redirect, and
+    a badge.
+"""
+
+from utils import warning, error, symlink, string_to_bool
+
+import sys
+import os
+import json
+
+################################################################################
+# Arguments and Paths
+num_args = len(sys.argv)
+
+if num_args < 1 or num_args > 4:
+    error('Unknown number of arguments')
+
+# relative paths to report directories from ROOT
+ROOT = 'data/reports'
+os.makedirs(ROOT, exist_ok = True)
+DIR_LAST_REPORT_REL = 'latest'
+OVERRIDE_LAST = True
+
+if num_args > 1: DIR_REPORT_REL = sys.argv[1]
+if num_args > 2: DIR_LAST_REPORT_REL = sys.argv[2]
+if num_args > 3: OVERRIDE_LAST = string_to_bool(sys.argv[3])
+
+DIR_REPORT = os.path.realpath(os.path.join(ROOT, DIR_REPORT_REL))
+DIR_LAST_REPORT_SYMBOLIC = os.path.join(ROOT, DIR_LAST_REPORT_REL)
+DIR_LAST_REPORT = os.path.realpath(DIR_LAST_REPORT_SYMBOLIC)
+
+REPORT_PATH = os.path.join(DIR_REPORT, 'test-status.json')
+LAST_REPORT_PATH = os.path.join(DIR_LAST_REPORT, 'test-status.json')
+
+if OVERRIDE_LAST:
+    DIR_BADGE = os.path.join('data/badges', DIR_LAST_REPORT_REL)
+    os.makedirs(DIR_BADGE, exist_ok = True)
+
+    DIR_REDIRECT = os.path.join('gh-pages', DIR_LAST_REPORT_REL)
+    os.makedirs(DIR_REDIRECT, exist_ok = True)
+
+################################################################################
+# Read current and previous test-status
+with open(REPORT_PATH, 'r') as f:
+    REPORT = json.load(f)
+
+if os.path.isfile(LAST_REPORT_PATH):
+    with open(LAST_REPORT_PATH, 'r') as f:
+        LAST_REPORT = json.load(f)
+else: # deal with the first run of this script
+    LAST_REPORT = {'pkgs': {}, 'hash': 'Unknown'}
+
+REPO = REPORT['repo']
+
+################################################################################
+# Generate report.md and test-status-diff.json
+REPORT_DIFF = {}
+REPORT_DIFF['current'] = REPORT_PATH.split('data/')[1]
+REPORT_DIFF['last'] = LAST_REPORT_PATH.split('data/')[1]
+REPORT_DIFF['total'] = REPORT['total']
+REPORT_DIFF['failure'] = REPORT['failure']
+REPORT_DIFF['success'] = REPORT['success']
+REPORT_DIFF['skipped'] = REPORT['skipped']
+
+with open(DIR_REPORT+'/report.md', 'w') as f:
+    # Header
+    f.write('# Package Evaluation Report\n\n')
+    f.write('## Job Properties\n\n')
+    f.write('*Testing:* [%s](%s) vs [%s](%s)\n\n' % (
+        REPORT_DIFF['current'].split('reports/')[1].split('/test-status.json')[0],
```

Comment:
This is rather dense code. What does this do? It seems to hardcode a lot of assumptions on what `REPORT_DIFF['current']` contains. What *does* it contain?  I am trying to untangle it... So we have these assignments:
```python
ROOT = 'data/reports'
...
DIR_LAST_REPORT_REL = 'latest'
if num_args > 2: DIR_LAST_REPORT_REL = sys.argv[2]
...
DIR_REPORT = os.path.realpath(os.path.join(ROOT, DIR_REPORT_REL))
REPORT_PATH = os.path.join(DIR_REPORT, 'test-status.json')
REPORT_DIFF['current'] = REPORT_PATH.split('data/')[1]
```
That's really complicated. Esp. once one factors the `realpath` in, because it means `DIR_REPORT` could be literally *anything* (and then I think the above line would produce garbage...?)

But let's suppose all is at default, no symlinks are involved, then we get:
```python
ROOT = 'data/reports'
DIR_LAST_REPORT_REL = 'latest'
DIR_REPORT = '/some/base/path/data/reports/latest'
REPORT_PATH = '/some/base/path/data/reports/latest/test-status.json'
REPORT_DIFF['current'] = 'reports/latest/test-status.json'
```
and so `REPORT_DIFF['current'].split('reports/')[1].split('/test-status.json')[0]` evaluates to *drumroll* `latest`

OK, so... Can't we turn this all on its header, and just have variables `report_id` and `last_report_id` (the latter defaulting to `latest`) or so. And then everything is computed and reported in terms of these. Like there could be a function `get_json_of_report()` which takes a "report id" and maps it to a path. And the line I am commenting at here would just print `last_report_id`. And so on.

Reply:
Yes it is quite messy at the moment. I will change it according to your suggestion.

Reply:
I think I did it this complicated way in order to deal with symlinks, so that i save the relative path `dir1/dir2/id` Instead of `latest`, since the symlink always changes. But I agree that it is too convoluted now.

Reply:
Ok, `latest` is indeed a virtual/symbolic id only, and dit can be useful to turn that into an "absolute" id. For that I'd suggest to store the "real" id into thE JSON file

Reply:
Ok, this is now simplified